[#661] Deduplicate trials via title

vitorbaptista opened the pull request:
This PR continues the work from https://github.com/opentrials/processors/pull/115 (I couldn't push to that branch, so needed to start a new one). Please check that for more information.

Before this is merged, we need to uncomment the `deploy` lines on `.travis.yml`

Connects opentrials/opentrials#661

arthurSena committed:
Initial approach for trial deduplication using scientific_title and brief_summary columns

arthurSena committed:
Refactor deduplication methods and increments deduplication log

arthurSena committed:
Merge branch 'master' into fix/{deduplicate_trials_via_title}

arthurSena committed:
Merge branch 'master' into fix/{deduplicate_trials_via_title}

arthurSena committed:
Alters deduplication method to use only scientific_title approach

arthurSena committed:
Merge branch 'opentrials-master'

arthurSena committed:
Merge branch 'master' into fix/{deduplicate_trials_via_title}

arthurSena committed:
Fixing some bugs and liting code

arthurSena committed:
Uses public_title approach to deduplicate trials that are from different sources

arthurSena committed:
Removes debug log

arthurSena committed:
Merge branch 'master' into fix/{deduplicate_trials_via_title}

arthurSena committed:
Linting code

arthurSena committed:
Adds find_trial method which encapsulates the logic of finding a trial via identifiers and public_title

arthurSena committed:
Adds audit log method which stores on database information about how the trial deduplication was made

arthurSena committed:
Fixes query on trial deduplication

vitorbaptista committed:
Merge commit '7fa4ad80ed90c31f2618c7a5c5e23f63741a81e6' into arthurSena/fix/{deduplicate_trials_via_title}

vitorbaptista committed:
[#661] Add SOURCE_COMMIT env and build containers on Travis

The SOURCE_COMMIT contains the source's commit hash, to be used inside the
container. We had to move the container building to Travis to have more
flexibility on how the "docker build" command is run, as Docker Cloud's
automatic builds don't allow us to pass extra variables (only Docker Hub).
Another great side effefct is that the containers will only be built if the
tests pass on Travis.

opentrials/opentrials#661

vitorbaptista committed:
Merge branch 'master' into arthurSena/fix/661-deduplicate-trials-via-title

vitorbaptista committed:
[#661] Remove "git" dependency, the commit hash comes from env

opentrials/opentrials#661

vitorbaptista committed:
Trying to fix Travis build

vitorbaptista committed:
[#661] Refactor trial/record writers to log deduplication

Deduplicating trials and logging the deduplication method required many
refactorings. The problem was that the parts of the code that searched for
duplicated trials, wrote trials, wrote records, and logged the deduplication
were scattered across the codebase.

With this commit, the process to write a trial became:

1. Call `.write_trial()` passing the trial's attributes (coming from the
warehouse)
  1. Search the records looking for a duplicated trial. If found, return it and
     how it was found (so we can log later).
2. Inside a transaction, do:
   1. Write the trial going over the same process we do now (checking if it's
      primary, etc.)
   2. Write the record
   3. Log the deduplication method. If it's a new trial, log with method
      "initial".

This means that `.write_trial()` is responsible for writing the trial, the
record, and logging the deduplication method. I don't like it much, as there're
many responsibilities in a single function. I don't like the name as well, as
`write_trial` is also writing a `record` and a `trial_deduplication_log`.

In hindsight, it seems it would've been better for all processors to simply
write `records`, not caring about `trials` at all, and then we would have a
dedicated processor that went over the `records` and created `trials`,
deduplicating as it went. I think this is a better separation of concerns.
Hopefully we'll get there.

opentrials/opentrials#661

vitorbaptista committed:
Merge branch 'master' into arthurSena/fix/661-deduplicate-trials-via-title

vitorbaptista committed:
Updating schemas

vitorbaptista commented:
Please, whoever review this check https://github.com/opentrials/api/pull/153 also

vitorbaptista committed:
Removing test commited by mistake

vitorbaptista commented:
Couldn't find the codacy report, but that shouldn't block merging this IMHO

vitorbaptista committed:
Fix codacity errors

victor-nitu commented on processors/base/helpers/__init__.py:
```diff
@@ -250,27 +250,94 @@ def find_trial_by_identifiers(conn, identifiers, ignore_record_id=None):
+        'SELECT trial_id FROM records'
+        ' WHERE public_title = :public_title'
+        ' AND source_id != :source_id'
+        ' AND id != :record_id'
```
As discussed, regarding `ignore_record_id` interpolation vs. conditional query building: it doesn't make any practical difference. Not sure about the readability being improved in any case, this is shorter, the other is more explicit. I suppose the only thing that matters is to be consistent in the two functions.

I'd add a short line comment to make things clearer tho, something like `# This avoids breaking the query when ignore_record_id is not supplied`

victor-nitu commented on processors/base/helpers/__init__.py:
```diff
@@ -250,27 +250,94 @@ def find_trial_by_identifiers(conn, identifiers, ignore_record_id=None):
+
+        if trial_found:
+            deduplication_method = 'public_title'
+
```
I would suggest:

* find by identifiers
  * if not found, find by public title
* if none of the above have results, try them again by using the `ignore_record_id`
* whenever anything is found, return (so we'd get rid of the misleading multiple indentation for code that has the same purpose)
  * alternatively, you could set a `trial_found` "flag variable" and run everything inside a `while` loop, or something similar – at the end of it, you just return it

However, this is not a crucial matter. It would probably increase the readability of code a bit.

victor-nitu commented on processors/base/writers/record.py:
```diff
@@ -70,25 +69,32 @@ def write_record(conn, record, source_id, trial_id, trial, is_primary):
+                conn['database']['records'].update(
+                    {
+                        'trial_id': record['trial_id'],
+                        'is_primary': False,
```
is this a 3 spaces indent?

vitorbaptista replied:
Looks fine to me :thinking:

victor-nitu commented on processors/base/writers/trial.py:
```diff
@@ -28,77 +32,106 @@ def write_trial(conn, trial, source_id, record_id):
+    latest_log = conn['database']['trial_deduplication_logs'].find_one(
+        trial_id=trial_id,
+        record_id=record_id,
+        order_by='-created_at'
```
I'll say it again: nice one :nerd_face:

victor-nitu reviewed (commented):
Hey, I wrote some comments on the diff. They're not blocking the merge, as the functionality isn't affected, but they might tackle readability a bit. Feel free to follow or discard them.

I haven't run it, so I'd say let's get it in today, together with the other PRs in the family, and start a big processing task over the weekend ;)

vitorbaptista committed:
Merge branch 'master' into arthurSena/fix/661-deduplicate-trials-via-title

vitorbaptista committed:
Update schemas

vitorbaptista committed:
Update schemas again, now with new API code

We didn't have the `trial_deduplication_logs` because it wasn't merged yet.
This is why the tests were failing.

vitorbaptista committed:
[#661] Try to ignore the record first when finding a trial

Before we tried to find it via identifiers ignoring the record, then without
ignoring it, and so on with the other methods (i.e. public_title). Instead of
that, it makes more sense to try every method ignoring the record, and only then
try the methods not ignoring it.

opentrials/opentrials#661

vitorbaptista committed:
Only deploy when on master

vitorbaptista committed:
[#661] .write_record() is now private to .write_trial()
